packages/expo-cli/src/commands/install.js

```diff
@@ -1,28 +1,129 @@
-      if (await Simulator.upgradeExpoAsync()) {
-        log('Done!');
+async function installAsync(packages, options) {
+  const projectRoot = findProjectRoot(process.cwd());
+  const { exp } = await ConfigUtils.readConfigJsonAsync(projectRoot);
+  const packageManager = PackageManager.createForProject(projectRoot, {
+    npm: options.npm,
+    yarn: options.yarn,
+  });
+
+  if (!fs.existsSync(path.join(exp.nodeModulesPath || projectRoot, 'node_modules'))) {
+    log.warn(`node_modules not found, running ${packageManager.name} install command.`);
+    await packageManager.installAsync();
+  }
+
+  const {
+    compatible,
+    incompatible,
+    dependencies,
+    sdkVersion,
+  } = await listBundledNativeModulesAsync(projectRoot, exp);
+
+  const nativeModules = [];
+  const others = [];
+  const versionedPackages = packages.map(arg => {
+    const spec = npmPackageArg(arg);
+    const { name } = spec;
+    if (
+      ['tag', 'version', 'range'].includes(spec.type) &&
+      name &&
+      (compatible[name] || incompatible[name])
+    ) {
+      if (compatible[name]) {
+        // Unimodule packages from npm registry are modified to use the bundled version.
+        const version = dependencies[name];
+        const modifiedSpec = `${name}@${version}`;
+        nativeModules.push(modifiedSpec);
+        return modifiedSpec;
+      } else {
+        throw new CommandError(
+          'INCOMPATIBLE_NATIVE_MODULE',
+          `'${name}' is incompatible with Expo SDK ${sdkVersion}.\nSupported SDK versions: ${
+            incompatible[name].sdkVersions
+          }`
+        );
       }
-    }, true);
+    } else {
+      // Other packages are passed through unmodified.
+      others.push(spec.raw);
+      return spec.raw;
+    }
+  });
+  const messages = [];
+  if (nativeModules.length > 0) {
+    messages.push(
+      `${nativeModules.length} SDK ${sdkVersion} compatible native ${inflect(
+        'modules',
+        nativeModules.length
+      )}`
+    );
+  }
+  if (others.length > 0) {
+    messages.push(`${others.length} other ${inflect('packages', others.length)}`);
+  }
+  log(`Installing ${messages.join(' and ')} using ${packageManager.name}.`);
+  await packageManager.addAsync(...versionedPackages);
+}
+
+function findProjectRoot(base) {
+  let previous = null;
+  let dir = base;
 
+  do {
+    if (fs.existsSync(path.join(dir, 'app.json'))) {
+      return dir;
+    }
+    previous = dir;
+    dir = path.dirname(dir);
+  } while (dir !== previous);
+
+  return base;
+}
+
+async function listBundledNativeModulesAsync(projectRoot, exp) {
+  const packageJsonPath = ConfigUtils.resolveModule('expo/package.json', projectRoot, exp);
+  if (!packageJsonPath) {
+    throw new CommandError(
+      'NO_EXPO',
+      "Package 'expo' not found in node_modules. Please make sure this is a managed Expo project."
+    );
+  }
+  const packageJson = await JsonFile.readAsync(packageJsonPath);
+  const sdkVersion = packageJson.version;
```

Comment:
is this reliable? version in package.json could be 32.0.3 for example. why not use app.json?

Reply:
The compatibility data only deals in major versions and we're using semver, so I think this should be reliable. See:
https://github.com/expo/expo-cli/blob/f96bd374d5415f192dfd0a7ee4bd7958954101a2/packages/xdl/src/modules/config.js#L44-L55

I went back and forth between using `app.json` version and using `package.json` version, but I think if use the word "SDK version" to mean the native version and not the `expo` SDK JS version (and I think we do), then reading the version from `app.json` is more correct. I'll change the code back to use `app.json`.